File: results/RQ1/05_baseline_qualification/units/p02_flappy_bird_game/java/accepted/repository/src/main/java/Game.java
/**
 * Orchestrates the Flappy Bird game state, managing the bird, pipes, scoring,
 * and game loop logic according to the specified rules.
 */
public class Game {
    /** The width of the game world */
    private final double worldWidth;
    
    /** The height of the game world */
    private final double worldHeight;
    
    /** The interval at which pipes are spawned */
    private final int spawnInterval;
    
    /** The fixed X position of the bird for collision and scoring */
    private final double birdX;
    
    /** The bird object representing the player */
    private Bird bird;
    
    /** List of active pipes in the game */
    private java.util.List<PipePair> pipes;
    
    /** Counter for the number of ticks elapsed */
    private long tickCount;
    
    /** Current score of the player */
    private int score;
    
    /** Current state of the game (RUNNING or GAME_OVER) */
    private GameState state;
    
    /** Random number generator for pipe placement */
    private RNG rng;

    /**
     * Creates a new game with the specified seed.
     * Delegates to Game(seed, 400.0, 600.0, 100).
     *
     * @param seed the seed for random number generation
     */
    public Game(long seed) {
        this(seed, 400.0, 600.0, 100);
    }

    /**
     * Creates a new game with the specified parameters.
     *
     * @param seed the seed for random number generation
     * @param worldWidth the width of the game world
     * @param worldHeight the height of the game world
     * @param spawnInterval the interval at which pipes are spawned
     */
    public Game(long seed, double worldWidth, double worldHeight, int spawnInterval) {
        this.worldWidth = worldWidth;
        this.worldHeight = worldHeight;
        this.spawnInterval = spawnInterval;
        this.birdX = 100.0; // Fixed bird X position as per specification
        
        reset(seed);
    }

    /**
     * Executes one tick of the game logic.
     * Returns false if the game is over, true otherwise.
     *
     * @return true if the game is still running, false if game over
     */
    public boolean tick() {
        // If game is over, do nothing
        if (state == GameState.GAME_OVER) {
            return false;
        }
        
        // Apply physics to the bird
        bird.applyGravity();
        bird.integrate();
        
        // Check for collisions with boundaries
        if (CollisionDetector.hitBounds(bird, 30.0, worldHeight)) {
            state = GameState.GAME_OVER;
            return false;
        }
        
        // Check for collisions with pipes
        for (PipePair pipe : pipes) {
            if (CollisionDetector.hitPipe(bird, 30.0, 30.0, pipe, worldHeight)) {
                state = GameState.GAME_OVER;
                return false;
            }
        }
        
        // Check for scoring
        for (PipePair pipe : pipes) {
            if (ScoreSystem.passedPipe(birdX, pipe)) {
                score++;
            }
        }
        
        // Increment tick count
        tickCount++;
        
        // Spawn new pipes if needed
        if (tickCount % spawnInterval == 0) {
            spawnPipe();
        }
        
        // Move all pipes
        for (PipePair pipe : pipes) {
            pipe.move();
        }
        
        // Remove off-screen pipes
        pipes.removeIf(pipe -> pipe.x() + pipe.width() < 0);
        
        return true;
    }

    /**
     * Resets the game state with a new seed.
     *
     * @param seed the new seed for random number generation
     */
    public void reset(long seed) {
        // Reset game state
        state = GameState.RUNNING;
        
        // Initialize RNG with new seed
        rng = new RNG(seed);
        
        // Reset game elements
        tickCount = 0;
        score = 0;
        pipes = new java.util.ArrayList<>();
        
        // Initialize bird at center of screen with zero velocity
        bird = new Bird(worldHeight / 2.0);
    }

    /**
     * Makes the bird flap if the game is running.
     */
    public void flapBird() {
        if (state == GameState.RUNNING) {
            bird.flap();
        }
    }

    /**
     * Gets the current Y position of the bird.
     *
     * @return the Y position of the bird
     */
    public double birdY() {
        return bird.y();
    }

    /**
     * Gets the current Y velocity of the bird.
     *
     * @return the Y velocity of the bird
     */
    public double birdVy() {
        return bird.vy();
    }

    /**
     * Gets the current tick count.
     *
     * @return the number of ticks elapsed
     */
    public long ticks() {
        return tickCount;
    }

    /**
     * Gets the current score.
     *
     * @return the current score
     */
    public int score() {
        return score;
    }

    /**
     * Gets the list of pipes.
     *
     * @return the list of active pipes
     */
    public java.util.List<PipePair> pipes() {
        return pipes;
    }
    
    /**
     * Spawns a new pipe at the right side of the screen.
     */
    private void spawnPipe() {
        // Generate random gap position
        double gapY = rng.nextDouble() * (worldHeight - 100) + 50;
        double gapHeight = 150.0; // Fixed gap height
        
        // Create new pipe pair
        PipePair pipe = new PipePair(
            worldWidth,   // Start at right edge of screen
            50.0,         // Fixed pipe width
            gapY,         // Gap top position
            gapHeight,    // Gap height
            3.0           // Fixed speed
        );
        
        pipes.add(pipe);
    }
}

File: results/RQ1/05_baseline_qualification/units/p02_flappy_bird_game/python/attempts/0002_seed_3131337434/p02_flappy_bird_game_python/p02_flappy_bird_game_python/flappy_bird/game.py
"""Game orchestration module for the Flappy Bird game."""

from typing import List, Tuple, Dict, Any
from flappy_bird.config import Config
from flappy_bird.physics import integrate_vy
from flappy_bird.scoring import is_pipe_scored
from flappy_bird.collision import hits_bounds, hits_pipe
from flappy_bird.pipes import Pipe


class Game:
    """Orchestration of the Flappy Bird game state machine and rules.

    This class manages the game's state, bird physics, pipe management,
    scoring, and collision detection according to the specified rules.
    """

    def __init__(self, config: Config, initial_pipes: List[Pipe] = None) -> None:
        """Initialize the game with a configuration and optional starting pipes.

        Args:
            config: Game configuration parameters.
            initial_pipes: Optional list of pipes to start with.
        """
        self._config = config
        self._state = "ready"  # 'ready', 'running', or 'game_over'
        self._tick = 0
        self._score = 0
        self._bird_y = config.start_y
        self._bird_vy = 0.0
        self._pipes = initial_pipes.copy() if initial_pipes else []

    def reset(self) -> None:
        """Reset the game to its initial state."""
        self._state = "ready"
        self._tick = 0
        self._score = 0
        self._bird_y = self._config.start_y
        self._bird_vy = 0.0
        self._pipes.clear()

    def flap(self) -> None:
        """Apply a flap impulse to the bird.

        In 'ready' state, this transitions to 'running' and applies the impulse.
        In 'running' state, re-applies the impulse.
        In 'game_over' state, does nothing.
        """
        if self._state == "game_over":
            return
        
        # Set state to running on first flap from ready
        if self._state == "ready":
            self._state = "running"
        
        # Apply the flap impulse (negative because up is negative y)
        self._bird_vy = self._config.flap_impulse

    def step(self) -> None:
        """Advance the game by one time step.

        Applies physics, updates positions, handles collisions, scoring, and
        pipe lifecycle in a deterministic order.
        """
        if self._state != "running":
            self._tick += 1
            return

        # 1. Integrate vertical velocity with gravity and clamp
        self._bird_vy = integrate_vy(
            self._bird_vy, self._config.gravity, self._config.max_vy
        )

        # 2. Update bird position
        self._bird_y += self._bird_vy

        # 3. Move all pipes
        for pipe in self._pipes:
            pipe.step(self._config.pipe_speed)

        # 4. Score pipes
        for pipe in self._pipes:
            if not pipe.passed:
                right_edge = pipe.right_edge()
                if is_pipe_scored(
                    self._config.bird_x, right_edge, pipe.passed
                ):
                    pipe.passed = True
                    self._score += 1

        # 5. Check collisions
        if (
            hits_bounds(self._bird_y, self._config.world_ceiling, self._config.world_floor)
            or any(
                hits_pipe(self._config.bird_x, self._bird_y, pipe)
                for pipe in self._pipes
            )
        ):
            self._state = "game_over"

        # 6. Remove off-screen pipes (after scoring)
        self._pipes = [
            pipe for pipe in self._pipes if pipe.right_edge() >= 0
        ]

        # 7. Increment tick
        self._tick += 1

    def pipes_readonly(self) -> Tuple[Pipe, ...]:
        """Get an immutable snapshot of the current pipes.

        Includes pipes that have been passed and are still in the list
        until they are fully off-screen and removed.

        Returns:
            A tuple of Pipe objects representing the current state of pipes.
        """
        return tuple(self._pipes)

    def status(self) -> Dict[str, Any]:
        """Get a snapshot of the current game state.

        Returns:
            A dictionary containing:
            - state: Current game state ('ready', 'running', 'game_over')
            - tick: Current tick count
            - y: Bird's vertical position
            - vy: Bird's vertical velocity
            - score: Current score
            - pipes: List of pipe dictionaries with their properties
        """
        return {
            "state": self._state,
            "tick": self._tick,
            "y": self._bird_y,
            "vy": self._bird_vy,
            "score": self._score,
            "pipes": [
                {
                    "x": pipe.x,
                    "gap_y": pipe.gap_y,
                    "width": pipe.width,
                    "gap_height": pipe.gap_height,
                    "passed": pipe.passed,
                }
                for pipe in self._pipes
            ],
        }